internal/junit/junit.go

```diff
@@ -72,3 +75,42 @@ func Parse(data []byte) (TestSuites, error) {
 
 	return tss, err
 }
+
+// GetFailedTests get failed test list from the JUnit report.
+func GetFailedTests(report TestSuites) []string {
+	classes := map[string]bool{}
+	for _, s := range report.TestSuites {
+		for _, tc := range s.TestCases {
+			if tc.Error != "" || tc.Failure != "" {
+				if tc.Name != "" {
+					classes[fmt.Sprintf("%s/%s", tc.ClassName, tc.Name)] = true
+				} else {
+					classes[tc.ClassName] = true
+				}
+			}
+		}
+	}
+	return getKeysFromMap(classes)
+}
+
+// GetFailedClasses get failed class list from the JUnit report.
+func GetFailedClasses(report TestSuites) []string {
+	classes := map[string]bool{}
+
+	for _, s := range report.TestSuites {
+		for _, tc := range s.TestCases {
+			if tc.Error != "" || tc.Failure != "" {
+				classes[tc.ClassName] = true
+			}
+		}
+	}
+	return getKeysFromMap(classes)
+}
+
+func getKeysFromMap(mp map[string]bool) []string {
+	var keys []string
+	for k := range mp {
+		keys = append(keys, k)
+	}
+	return keys
```

Comment:
Not sure how performance relevant this piece here is, but we know the size of the map already, so we can pre-allocate the slice instead of having it dynamically expand as we add values to it.

Reply:
The slice increasing speed is quite fast so I don't think it should be an issue. But I can update it to the fixed length way